scm-core/src/test/java/sonia/scm/plugin/SmpArchiveTest.java

```diff
@@ -157,10 +155,7 @@ private void writeDescriptor(File descriptor, String name, String version) throw
       writer.writeEndElement();
       writer.writeEndElement();
       writer.writeEndDocument();
-    } finally {
-      if (writer != null) {
-        writer.close();
-      }
```

Comment:
I think this null check is still useful, because, if for example `XMLOutputFactory.newFactory()` throws an exception, then `writer` would be null, `writer.close()` would throw a NullPointerException, and you would never know what really happened.

Reply:
@pfeuffer if exception will be thrown - code execution will never hit `writer.close()`

Reply:
@pfeuffer in this case there is used class - XMLStreamWriterImpl.
Which on close() flushes underlaying resource.
So after try-catch-with-resource, corresponding resource will be already closed, and any attempt to flush it will lead to exception.
So I just put close inside try-catch block.
And that works fine, but only while JRE uses the same bad apache xerces library.
If on JRE 11 --> 777 migration JRE will start to use another built-in xml serialization library, which will start to close resources - another problem will appear.
Ideally, I would suggest to forget about runtime linking to unpredictable serialization libraries in favor of FasterXML's Jackson library, which can do everything and even more. (In serialization/deserialization context of XML/json)

Reply:
@StNekroman regarding the NPE, take this little example. This will throw a NPE, because the `finally` block will be run even in case of an exception. You will not see the `IllegalStateException`.

```java
public class SomeFinallyTest {

  public static void main(String[] args) throws IOException {
    String x = null;
    try (OutputStream stream = createStream()) {
      x = "Got it";
    } finally {
      System.out.println(x.toLowerCase());
    }
  }

  static OutputStream createStream() {
    throw new IllegalStateException("just for demonstration");
  }
}
```

We will definitely talk about the serialization library. Thanks for the hint!

Reply:
@StNekroman regarding the NPE, take this little example. This will throw a NPE, because the `finally` block will be run even in case of an exception. You will not see the `IllegalStateException`.

```java
public class SomeFinallyTest {

  public static void main(String[] args) throws IOException {
    String x = null;
    try (OutputStream stream = createStream()) {
      x = "Got it";
    } finally {
      System.out.println(x.toLowerCase());
    }
  }

  static OutputStream createStream() {
    throw new IllegalStateException("just for demonstration");
  }
}
```

We will definitely talk about the serialization library. Thanks for the hint!

Reply:
@pfeuffer yes, but... there is no finally block at all!